src/cli/rustup_mode.rs

```diff
@@ -851,9 +859,31 @@ fn target_list(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
 fn target_add(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
     let toolchain = explicit_or_dir_toolchain(cfg, m)?;
 
-    for target in m.values_of("target").expect("") {
-        let new_component = Component::new("rust-std".to_string(), Some(TargetTriple::new(target)));
+    let mut targets: Vec<String> = m
+        .values_of("target")
+        .expect("")
+        .map(ToString::to_string)
+        .collect();
+
+    if targets.contains(&"all".to_string()) {
+        targets.clear();
```

Comment:
Is it worth emitting a warning here if `targets.len() != 1` ?

Reply:
I'm not sure, I also asked myself this, and did that initially. That way: `rustup target add x86_64-unknown-linux-gnu all` errors. 

But then I thought, what does that cmd actually mean? It reads "install target `x86_64-unknown-linux-gnu` and also all available targets". This is a weird thing to request, as in, the user could just have requested installing all available targets, and that would have the same effect. 

But that's the only reason I got for "why should that error". I mean, the request isn't wrong, and it is easily satisfiable, so what's the value of erroring there?

Reply:
So at the end I just changed to accept it, and that simplified the code a bit (no need for an error message here, etc.)

Reply:
My gut feeling is that the invocation should be **either** `rustup target add foo-bar-gnu...` **or** `rustup target add all` but if noone else backs me up then I'm okay to back down as it were.

Reply:
I'm fine either way :) I used to just have an `assert!` here, to change this into an error, is there a different API that I should use ?

Reply:
The function returns a `Result<()>` which comes from our error chain module I believe.  You could add a `TargetAllSpecifiedWithTargets` variant to `src/cli/errors.rs` and return that which ought to do the right thing.  Remember at that point to add a test to assert it happens :D

Reply:
The last commit should resolve this - thank you for the hints.